### PR TITLE
Fix causal LM loss alignment and optimize evaluation memory usage

### PR DESCRIPTION
**Pull request by Eamon2009**
### Overview

This PR improves the **base model evaluation pipeline** by fixing logical issues, improving reliability, and enhancing performance for both **nanochat models** and **HuggingFace models**.

The changes focus on **correct evaluation logic, distributed safety, and better engineering practices** for large language model evaluation.

---

### 1. Correct Causal Language Modeling Loss

Fixed the loss computation to correctly align logits and targets for causal language models.

**Before**

* Loss computed directly between logits and targets.
* Caused misaligned token prediction and incorrect evaluation metrics.

**After**

* Implemented proper token shifting so token *n* predicts token *n+1*.
* Ensures correct loss and BPB calculation.

---

### 2. Memory-Efficient Evaluation

Wrapped evaluation logic with `torch.no_grad()`.

Benefits:

* Prevents gradient tracking
* Reduces GPU memory usage
* Improves evaluation speed

Applied to:
* CORE benchmark evaluation
* BPB evaluation
* Sampling

---

### 3. Robust Sequence Length Detection

Improved HuggingFace model sequence length detection.

**Before**

```python
max_seq_len = 1024 if "gpt2" in hf_path else None
```

**After**

```python
max_seq_len = getattr(model.config, "max_position_embeddings", 1024)
```

This now works correctly for:

* GPT-2
* LLaMA
* Mistral
* Falcon
* other HuggingFace models

---

### 4. Token Byte Computation Caching

Token byte computation is expensive because it decodes every token in the vocabulary.

Added caching mechanism:

* Saves computed tensor to disk
* Avoids recomputation on subsequent runs

Benefits:

* Faster startup
* Reduced evaluation overhead

---

### 5. Distributed Evaluation Safety

Improved behavior in multi-GPU environments.

Changes:

* CORE benchmark evaluation runs only on **rank 0**
* Prevents redundant computation across GPUs

Benefits:

* Saves compute resources
* Improves distributed efficiency

---

### 6. Safer BPB Evaluation Step Calculation

Prevented edge cases where evaluation steps could become zero.

**Before**

```python
steps = args.split_tokens // tokens_per_step
```

**After**

```python
steps = max(1, args.split_tokens // tokens_per_step)
```

Ensures evaluation always performs at least one step.

---

### 7. Cleaner Sampling Logic

Improved sampling structure:

* Centralized sampling logic
* Ensured inference runs under `torch.no_grad()`
* Reduced duplicated code

---

### 8. General Code Quality Improvements

Additional improvements include:

* Clearer function structure
* Improved logging messages
* Safer model loading logic
* More consistent variable naming
* Cleaner evaluation flow

---

## Result

The updated script provides:

* **Correct evaluation metrics**
* **Better GPU memory efficiency**
* **Faster evaluation runs**
* **Improved distributed execution**
* **Better compatibility with HuggingFace models**

Overall, this makes the evaluation pipeline **more robust and production-ready** for large language model benchmarking.

### CODE REVIEW

**Comment by Eamon2009** [comment]
Thanks for reviewing the PR.

You're right — the diff includes formatting and reordering which makes the review harder.

Regarding the implementation: parts of the  draft were assisted by AI, but I reviewed the code line by line and verified the logic myself. I’m still exploring and learning best practices, so feedback like this is very helpful.

Testing:

Ran training on the tinyshakespeare dataset

Verified the training loop runs without errors

No regressions observed in this small test

Due to limited GPU resources, the test was run on a very small GPU with a reduced dataset and short training run, mainly to ensure correctness rather than benchmark performance.

**Comment by Eamon2009** [comment]
Thanks for the feedback! 
I realize now that this PR is too bloated and the diff is quite noisy due to the formatting changes.
I definitely respect your time and want to make this easier to review. I’m going to close this PR and focused PRs later.

Thanks for pointing me in the right direction!
